File: 938. Range Sum of BST.py
'''
Given the root node of a binary search tree and two integers low and high, return the sum of values of all nodes with a value in the inclusive range [low, high].
'''
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def rangeSumBST(self, root, low, high):
        """
        :type root: TreeNode
        :type low: int
        :type high: int
        :rtype: int
        """
        if not root:
            return 0
        #print(root.val)
        n = 0
        if low <= root.val <= high:
            n += root.val
        if root.val < high:
            n += self.rangeSumBST(root.right, low, high)
        if root.val > low:
            n += self.rangeSumBST(root.left, low, high)
        
        return n

File: 938. Range Sum of BST.c
/*
Given the root node of a binary search tree and two integers low and high, return the sum of values of all nodes with a value in the inclusive range [low, high].
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int rangeSumBST(struct TreeNode* root, int low, int high) {
    if(root == NULL){
        return 0;
    }
    int total = 0;
    if(root->val <= high && root->val >= low){// in range, add and search left+right
        total += root->val;
        total += rangeSumBST(root->right, low, high);
        total += rangeSumBST(root->left, low, high);
    }else if(root->val < low){// search right
        total += rangeSumBST(root->right, low, high);
    }else {// search left
        total += rangeSumBST(root->left, low, high);
    }
    return total;
}